### PR TITLE
Fix unit of measurement fields for DSMR power consumed/delivered fields

### LINKED ISSUE #2353: DSMR component using W instead of kW
### The problem

DSMR uses kW for all its values, see for example this small excerpt from the specs:

afbeelding

However, the DSMR component uses "W" as the unit type.

### Which version of ESPHome has the issue?

2021.8.0

### What type of installation are you using?

Docker

### Which version of Home Assistant has the issue?

2021.8.0

### What platform are you using?

ESP8266

### Board

ESP-01

### Component causing the issue

dsmr

### Example YAML snippet

_No response_

### Anything in the logs that might be useful for us?

_No response_

### Additional information

This was raised as an issue in the comments for the DSMR PR:
https://github.com/esphome/esphome/pull/1881#discussion_r667531234
Frenck requested to post this as a separate issue.

### PR DESCRIPTION
**Pull request by jsuanet**
# What does this implement/fix? 

As described in issue #2353 the DSMR component still uses the wrong unit of measurement (W) instead of kW for the power consumed/delivered fields. Also some fields use string values for the units instead of the already/new defined UNIT_* constants in const.py file. This PR fixes issue #2353.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/2353

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [X] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/dsmr/sensor.py [hunk 1]
```diff
@@ -26,40 +30,80 @@
...
         cv.Optional("power_returned"): sensor.sensor_schema(
-            UNIT_WATT, ICON_EMPTY, 3, DEVICE_CLASS_POWER, STATE_CLASS_MEASUREMENT
+            UNIT_KILOWATT, ICON_EMPTY, 3, DEVICE_CLASS_POWER, STATE_CLASS_MEASUREMENT
         ),
         cv.Optional("reactive_power_delivered"): sensor.sensor_schema(
-            "kvar", ICON_EMPTY, 3, DEVICE_CLASS_ENERGY, STATE_CLASS_NONE
+            UNIT_KILOVOLT_AMPS_REACTIVE,
+            ICON_EMPTY,
+            3,
+            DEVICE_CLASS_ENERGY,
```
Comment (oxan) [inline esphome/components/dsmr/sensor.py]: Shouldn't this be `DEVICE_CLASS_POWER`, as already the case for the individual phases?

File: esphome/components/dsmr/sensor.py [hunk 2]
```diff
@@ -26,40 +30,80 @@
...
+            STATE_CLASS_NONE,
```
Comment (oxan) [inline esphome/components/dsmr/sensor.py]: And this `STATE_CLASS_MEASUREMENT`, as is already the case for reactive power returned and the individual phases?

File: esphome/components/dsmr/sensor.py [hunk 3]
```diff
@@ -26,40 +30,80 @@
...
         ),
         cv.Optional("reactive_power_returned"): sensor.sensor_schema(
-            "kvar", ICON_EMPTY, 3, DEVICE_CLASS_ENERGY, STATE_CLASS_MEASUREMENT
+            UNIT_KILOVOLT_AMPS_REACTIVE,
+            ICON_EMPTY,
+            3,
+            DEVICE_CLASS_ENERGY,
```
Comment (oxan) [inline esphome/components/dsmr/sensor.py]: Same here, should this be `DEVICE_CLASS_POWER`?

File: esphome/components/dsmr/sensor.py [hunk 4]
```diff
@@ -101,40 +145,40 @@
...
-            UNIT_WATT, ICON_EMPTY, 3, DEVICE_CLASS_POWER, STATE_CLASS_MEASUREMENT
+            UNIT_KILOWATT, ICON_EMPTY, 3, DEVICE_CLASS_POWER, STATE_CLASS_MEASUREMENT
         ),
         cv.Optional("power_returned_l3"): sensor.sensor_schema(
-            UNIT_WATT, ICON_EMPTY, 3, DEVICE_CLASS_POWER, STATE_CLASS_MEASUREMENT
+            UNIT_KILOWATT, ICON_EMPTY, 3, DEVICE_CLASS_POWER, STATE_CLASS_MEASUREMENT
         ),
         cv.Optional("reactive_power_delivered_l1"): sensor.sensor_schema(
-            UNIT_WATT, ICON_EMPTY, 3, DEVICE_CLASS_POWER, STATE_CLASS_MEASUREMENT
+            UNIT_KILOWATT, ICON_EMPTY, 3, DEVICE_CLASS_POWER, STATE_CLASS_MEASUREMENT
```
Comment (oxan) [inline esphome/components/dsmr/sensor.py]: I think this should be `UNIT_KILOVOLT_AMPS_REACTIVE` instead (same for the other 5 below).

**Review by oxan — COMMENTED** [review]
LGTM.

Tangentially related, but I also noticed some inconsistencies. Don't let this delay the PR, but might be nice to tackle them as well while we're at it.

File: esphome/components/dsmr/sensor.py (hunk 1 above)
Comment (jsuanet) [reply esphome/components/dsmr/sensor.py]: You are right, I will check them all

**Review by oxan — APPROVED** [review]

**Review by jesserockz — APPROVED** [review]
@glmnet or @zuidwijk could you please verify this PR?

**Comment by zuidwijk** [comment]
https://www.netbeheernederland.nl/_upload/Files/Slimme_meter_15_a727fce1f1.pdf

Specifications states:

Actual electricity power delivered (+P) in 1 Watt resolution
Actual electricity power received (-P) in 1 Watt resolution

That is current power either delivered from, or returned to the grid.

**Comment by zuidwijk** [comment]
image
Spec says Watt resolution. My meter (5.0) gives kW format.

**Comment by mmakaay** [comment]
Correct, 1 W resolution, which is accomplished by 0.001 kW steps on the meter.
Tested the PR and it looks good to me too.

**Pull request merged**